compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -301,33 +301,30 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         obligation: &TraitObligation<'tcx>,
     ) -> SelectionResult<'tcx, Selection<'tcx>> {
         let candidate = match self.select_from_obligation(obligation) {
-            Err(SelectionError::Overflow(OverflowError::Canonical)) => {
+            SelectionResult::Error(SelectionError::Overflow(OverflowError::Canonical)) => {
                 // In standard mode, overflow must have been caught and reported
                 // earlier.
                 assert!(self.query_mode == TraitQueryMode::Canonical);
-                return Err(SelectionError::Overflow(OverflowError::Canonical));
-            }
-            Err(SelectionError::Ambiguous(_)) => {
-                return Ok(None);
+                return SelectionResult::Error(SelectionError::Overflow(OverflowError::Canonical));
             }
-            Err(e) => {
-                return Err(e);
+            SelectionResult::Error(SelectionError::Ambiguous(_)) | SelectionResult::Ambiguous => {
```

Comment:
why does `SelectionError::Ambiguous` exist? That's really weird.

It seems like the only place where that's emitted is `fn candidate_from_obligation_on_cache` which should be able to just return `Ok(None)`/ `SelectionResult::Ambiguous`?

Reply:
It also returns the set of impl candidates that it was not able to winnow down, for error reporting purposes. We could probably return `SelectionResult::Ambiguous`, but then we'd need some other way of asking selcx what impl candidates we should display as possibly applying.

Reply:
> but then we'd need some other way of asking selcx what impl candidates we should display as possibly applying.

yes, i think that's what we should do here. HAven't looked into how difficult this is, but it definitely feels feasible to me